R/fragments-calculate2.R

```diff
@@ -0,0 +1,279 @@
+        df[[i]] <- data.frame(mz = c(amz_mod[[i]], cmz_mod[[i]]),
+                              ion = c(aion, cion),
+                              type = c(atype, ctype),
+                              pos = pos,
+                              z = z,
+                              seq = c(aseq, cseq),
+                              stringsAsFactors = FALSE)
+        df[[i]] <- .neutralLoss(df[[i]],
+                                water = neutralLoss$water,
+                                ammonia = neutralLoss$ammonia)
+        df[[i]] <- .terminalModifications(df[[i]],
+                                          modifications = fixed_modifications)
+        rownames(df[[i]]) <- NULL
+        non_zero <- mod_combinations[[i]] != 0
+        names(mod_combinations[[i]])[non_zero] <- 
+            paste0("[", (names(mod_combinations[[i]])[non_zero]), "]")
+        df[[i]][["peptide"]] <- paste(names(mod_combinations[[i]]),
+                                      collapse = "")
+    }
+    
+    df <- do.call(rbind, df)
+    rownames(df) <- NULL
+    df
+}
+
+#' @title Generates list of possible combinations of modifications
+#' 
+#' @param sequence Character. A peptide sequence that may have modifications or not 
+#' 
+#' @param fixed_modifications Named numeric. Specifies which fixed modifications are used
+#' 
+#' @param variable_modifications Named numeric. Specifies which variable modifications are used
+#' 
+#' @param max_mods Numeric. Indicates how many modifications can be applied at once.
+#' 
+#' @return list with all possible combinations of modifications
+#' 
+#' @noRd
+#' 
+#' @examples
+#' .modificationPositions("ARGHKA", variable_modifications = c(A = 4, K = 5, S = 8), max_mods = 3)
+#' 
+.modificationPositions <- function(sequence,
+                                   variable_modifications = NULL,
+                                   max_mods = Inf) {
+    
+    sequence_split <- strsplit(sequence, "")[[1]]
+    
+    if (length(variable_modifications)) {
+        modifiable_positions_var <-
+            which(sequence_split %in% names(variable_modifications))
+        l <- length(modifiable_positions_var)
+    } else {l <- NULL}
+    
+    ## take the maximum amount of modifications possible 
+    max_mods <- min(max_mods, l)
+    
+    if (max_mods > 0 & length(l)) {
+        mod_combinations <- lapply(0:max_mods, function(n) {
+            if (n == 0) {
+                list(integer(0))
+            } else if (l == 1) {
+                list(modifiable_positions_var)
+            } else {
+                combn(modifiable_positions_var, n, simplify = FALSE)
+            }
+        })
+        
+        mod_combinations <- do.call(c, mod_combinations)
+        lapply(mod_combinations, function(comb) {
+            mods <- rep(0, nchar(sequence))
+            mods[comb] <- unname(variable_modifications[sequence_split[comb]])
+            names(mods) <- sequence_split
+            mods
+        })
```

Comment:
This second `lapply` loop is unnecessary. You are already looping to the same combinations in the `lapply` call above. My suggestion for the complete function would be something like (not tested for all edge-cases):
```r
.modificationPositions <- function(sequence,
                                   variable_modifications = NULL,
                                   max_mods = Inf) {
    sequence_split <- strsplit(sequence, "", fixed = TRUE)[[1]]

    modifiable_positions_var <-
        which(sequence_split %in% names(variable_modifications))

    l <- length(modifiable_positions_var)

    ## take the maximum amount of modifications possible
    max_mods <- min(max_mods, l)

    if (!length(variable_modifications) || max_mods <= 0)
        return(
            list(setNames(integer(length(sequence_split)), sequence_split))
        )

    .mod <- function(cmb,
                     seq_split = sequence_split,
                     var_mods = variable_modifications) {
        m <- setNames(integer(length(seq_split)), seq_split)
        m[cmb] <- var_mods[seq_split[cmb]]
        m
    }

    c(
        list(setNames(integer(length(sequence_split)), sequence_split)),
        if (length(modifiable_positions_var) == 1)
            lapply(modifiable_positions_var, .mod)
        else
            unlist(
                lapply(seq_len(max_mods),
                    function(n)combn(
                        modifiable_positions_var, n,
                        FUN = .mod,
                        simplify = FALSE
                    )
                ),
                recursive = FALSE
            )
    )
}
```